### PR TITLE
Faz implementation [WIP]

### PR DESCRIPTION
**Pull request by wsuppiger**
The make purpose of this PR is to create the link between Caw Functions and the Faz backend.  Also, I know I should make everything in CawFunction static, I just didn't have time in this PR.

### CODE REVIEW

**Comment by wsuppiger** [comment]
I actually might not change the Caw Function setup.  Please let me know what you think

File: src/faz/faz_server.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  if (caw_method == nullptr) {
+    LOG(WARNING) << "not found in CawFunction class for function "
+                 << event_function;
+    return Status(StatusCode::NOT_FOUND, "function call not found");
+  }
+  CawFuncReply func_reply = (caw_.*(caw_method))(request->payload());
+
+  Any* any = new Any();
+  switch (event_type) {
+    case 1:  // registeruser
```
Comment (barath) [inline src/faz/faz_server.cc]: Any values in a switch should be declared in an enum or similar and used here via the enum declarations.

File: src/faz/faz_server.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      break;
+    case 2:  // caw
+      CawReply caw_reply;
+      caw_reply.ParseFromString(func_reply.message);
+      any->PackFrom(caw_reply);
+      break;
```
Comment (barath) [inline src/faz/faz_server.cc]: You should probably put TODOs in for the other event types.

File: src/faz/faz_server.cc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                 << event_function;
+    return Status(StatusCode::NOT_FOUND, "function call not found");
+  }
+  CawFuncReply func_reply = (caw_.*(caw_method))(request->payload());
+
+  Any* any = new Any();
+  switch (event_type) {
+    case 1:  // registeruser
+      break;
+    case 2:  // caw
```
Comment (barath) [inline src/faz/faz_server.cc]: Why not just an if statement if there's just one?

File: src/caw/caw_function.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+class CawFunction {
+  typedef CawFuncReply (CawFunction::*caw_method_t)(const Any& payload);
+  typedef std::unordered_map<std::string, caw_method_t> caw_func_map_t;
+
+ public:
+  explicit CawFunction(KeyValueClient& kv)
+      : kv_(kv),
+        function_map_{{"registeruser", &CawFunction::RegisterUser}}
+
```
Comment (barath) [inline src/caw/caw_function.h:45]: Remove blank line

File: src/caw/caw_function.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  {}
+
+  // name of function to function pointer map
+  caw_method_t GetFunction(const std::string);
```
Comment (barath) [inline src/caw/caw_function.h]: Reference

File: src/caw/caw_function.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  // register user to kv storage
+  CawFuncReply RegisterUser(const Any& payload);
```
Comment (barath) [inline src/caw/caw_function.h]: This and all functions like it should be static or outside of the class (in a namespace)

File: src/caw/caw_function.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+ private:
+  // client to communicate with key value server
+  KeyValueClient& kv_;
```
Comment (barath) [inline src/caw/caw_function.h]: You'll have to pass this into the functions so they can use it

File: src/caw/caw_function.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+using caw::CawReply;
+using caw::FollowRequest;
+using caw::FollowReply;
+using caw::ReadRequest;
+using caw::ReadReply;
+using caw::ProfileRequest;
+using caw::ProfileReply;
+
+// struct for caw function reply message
+struct CawFuncReply {
```
Comment (barath) [inline src/caw/caw_function.h:36]: Why is this needed?  Can't you use the protobuf?

Comment (wsuppiger) [reply src/caw/caw_function.h:36]: I am passing both the reply status and serialized Reply message back through faz

File: src/faz/faz_server.cc (hunk 1 above)
Comment (wsuppiger) [reply src/faz/faz_server.cc]: fixed in next PR

File: src/faz/faz_server.cc (hunk 3 above)
Comment (wsuppiger) [reply src/faz/faz_server.cc]: fixed in next PR

File: src/faz/faz_server.cc (hunk 2 above)
Comment (wsuppiger) [reply src/faz/faz_server.cc]: fixed in next PR

File: src/caw/caw_function.h (hunk 4 above)
Comment (wsuppiger) [reply src/caw/caw_function.h:45]: fixed in next PR

File: src/caw/caw_function.h (hunk 7 above)
Comment (wsuppiger) [reply src/caw/caw_function.h]: can I make this static to fix it?

File: src/caw/caw_function.h (hunk 5 above)
Comment (wsuppiger) [reply src/caw/caw_function.h]: fixed in next PR

File: src/caw/caw_function.h (hunk 6 above)
Comment (wsuppiger) [reply src/caw/caw_function.h]: Resolved in next PR

File: src/caw/caw_function.h (hunk 7 above)
Comment (wsuppiger) [reply src/caw/caw_function.h]: removed

File: src/caw/caw_function.h (hunk 8 above)
Comment (barath) [reply src/caw/caw_function.h:36]: Hmm, I'm still a bit confused -- why isn't GRPC Status sufficient for the status?  You should be looking at the status (via GRPC) as separate from the reply message (which should be just the caw message, unless I'm missing something), but both come back from the server in the GRPC response.

Comment (wsuppiger) [reply src/caw/caw_function.h:36]: Caw, Read, and Profile all return serialized information, so I guess a better name for `CawFuncReply.message` would be `CawFuncReply.serialized_obj`.  GRPC Status allows you to return a status with a status message, but I did not want to store the serialized object in status message since I needed to pack it in an EventReply (protobuf Any) in Faz server.
